src/intrusive_pairing_heap.rs

```diff
@@ -0,0 +1,402 @@
+        if let Some(mut parent) = parent {
+            // Unlink this node from its parent.
+            if let Some(mut prev) = node.prev {
+                prev.as_mut().next = node.next;
+            } else {
+                parent.as_mut().first_child = node.next;
+            }
+            if let Some(mut next) = node.next {
+                next.as_mut().prev = node.prev;
+            }
+            node.next = None;
+            node.prev = None;
+        } else {
+            debug_assert_eq!(node.next, None);
+            debug_assert_eq!(node.prev, None);
+            debug_assert_eq!(self.root, Some(node.into()));
+            self.root = None;
+        }
+        if let Some(first_child) = node.first_child.take() {
+            // Merge together the children.
+            let children = merge_children(first_child);
+            // Add the children back into the parent.
+            if let Some(parent) = parent {
+                // The heap property is preserved because we had `parent.data`
+                // <= `node.data`, and `node.data` <= `child.data` for all
+                // children.
+                add_child(parent, children);
+            } else {
+                self.root = Some(children);
+            }
+        }
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
+    use super::{HeapNode, PairingHeap};
+    use core::ptr::NonNull;
+
+    // Recursively check the provided node and all descendants for:
+    // - pointer consistency: parent pointers and next/prev
+    // - the heap property: `node.data <= child.data` for all children
+    unsafe fn validate_heap_node<T: Ord>(
+        node: &HeapNode<T>,
+        parent: Option<&HeapNode<T>>,
+    ) {
+        assert_eq!(node.parent, parent.map(NonNull::from));
+        if let Some(p) = parent {
+            assert!(p.data <= node.data);
+        }
+        if let Some(prev) = node.prev {
+            assert_eq!(prev.as_ref().next, Some(node.into()));
+        }
+        if let Some(next) = node.next {
+            assert_eq!(next.as_ref().prev, Some(node.into()));
+        }
+        let mut child = node.first_child;
+        while let Some(c) = child {
+            validate_heap_node(c.as_ref(), Some(node));
+            child = c.as_ref().next;
+        }
+    }
+
+    fn validate_heap<T: Ord>(heap: &PairingHeap<T>) {
+        if let Some(root) = heap.root {
+            // This is also sufficient to check that `heap.root` is indeed a
+            // minimum element of the heap.
+            unsafe {
+                validate_heap_node(root.as_ref(), None);
+            }
+        }
+    }
+
+    #[test]
+    fn insert_and_remove() {
+        // This test exhaustively covers every possible schedule of inserting,
+        // then removing, each of five different nodes from the heap.
+        #[derive(Copy, Clone, Debug)]
+        enum Action {
+            Insert(u8),
+            Remove(u8),
+        }
+        fn generate_schedules(
```

Comment:
Nit: An empty line before/after function/type declarations makes things easier to read